Adjust BSD os::thread_cpu_time() implementation:

bsdkurt opened the pull request:
* For BSD's that have getrusage(RUSAGE_THREAD), prefer its use over
  pthread_getcpuclockid(3) since it can support both user and system cpu
  time.
* On OpenBSD, stop using pthread_getcpuclockid(3) and provide new
  implementation that supports both user and system cpu time and doesn't
  crash the JVM when called on a thread that has exited.
* Remove unnecessary checks for Bsd::_getcpuclockid != NULL in various
  methods since the underlying implementation checks for it as well as
  os::is_thread_cpu_time_supported().

Notes:
On OpenBSD:
```
gmake run-test-only TEST="runtime/jni/terminatedThread/TestTerminatedThread.java"
```
Crashed the jvm by calling pthread_getcpuclockid on a terminated thread. It appears there isn't a safe way to check if a thread is terminated in the JVM and the pthread_getcpuclockid implementation doesn't work when the caller wants just user time anyway so I coded up a alternate solution.

Making getrusage(RUSAGE_THREAD) preferred for *BSD for the the current thread wasn't strictly necessary but I think makes sense if FreeBSD and NetBSD provide fully functional implementations similar to what I have done.

battleblow commented on src/hotspot/os/bsd/os_bsd.cpp:
```diff
@@ -4024,6 +4006,61 @@ jlong os::thread_cpu_time(Thread *thread, bool user_sys_cpu_time) {
+    if (getrusage(RUSAGE_THREAD, &usage) != 0)
+      return -1;
+    nanos = (jlong)usage.ru_utime.tv_sec * NANOSECS_PER_SEC;
+    nanos += (jlong)usage.ru_utime.tv_usec * 1000;
```
Maybe a `NANOSECS_PER_MICROSEC` constant instead of hard coding 1000.

battleblow commented on src/hotspot/os/bsd/os_bsd.cpp:
```diff
@@ -4024,6 +4006,61 @@ jlong os::thread_cpu_time(Thread *thread, bool user_sys_cpu_time) {
+  pid_t pid = getpid();
+  struct kinfo_proc *ki;
+
+  int mib[] = { CTL_KERN, KERN_PROC, KERN_PROC_PID|KERN_PROC_SHOW_THREADS, pid, sizeof(struct kinfo_proc), 0 };
```
Reading the manual pages, I think this would work on FreeBSD with `KERN_PROC_INC_THREAD` rather than `KERN_PROC_SHOW_THREADS`.  I can test that afterwards though.

battleblow commented on src/hotspot/os/bsd/os_bsd.cpp:
```diff
@@ -4024,6 +4006,61 @@ jlong os::thread_cpu_time(Thread *thread, bool user_sys_cpu_time) {
   }
 #else
+#ifdef RUSAGE_THREAD
+  if (thread == Thread::current()) {
```
I'm wondering how much value having this special case provides.  Is the thread being used often the current thread or does this execute rarely?  Can this thread be preempted between here and the call to getrusage?

bsdkurt replied:
I'm assuming that if a thread queries its own cpu usage it would hit the RUSAGE_THREAD implementation. I kept it for OpenBSD since the getrusage implementation is a single syscall vs two syscalls and malloc/free. I have no strong feelings on keeping it or just leaving it for the clockgettime implementation (where it adds a bit of functionality). Probably an unnecessary micro optimization. If you have a preference let me know.

battleblow reviewed (commented):
I had some comments/questions.  But they aren't major, so no objections to pushing it.

bsdkurt committed:
Adjust BSD os::thread_cpu_time() implementation
* On OpenBSD, stop using pthread_getcpuclockid(3) and provide new
  implementation that supports both user and system cpu time and doesn't
  crash the JVM when called on a thread that has exited.
* Remove unnecessary checks for Bsd::_getcpuclockid != NULL in various
  methods since the underlying implementation checks for it as well as
  os::is_thread_cpu_time_supported().

bsdkurt commented:
@battleblow I decided to not make the change to RUSAGE_THREAD and leave it the way it was. NetBSD doesn't have RUSAGE_THREAD, so if you convert the FreeBSD implementation over to  KERN_PROC_PID w/threads as well it can be removed then.

Here are the relevant tests for testing:
```
gmake run-test-only TEST="hotspot/jtreg/runtime/jni/terminatedThread/TestTerminatedThread.java jdk/jfr/event/runtime/TestThreadCpuTimeEvent.java"
```
okay to merge the current version?

battleblow commented:
> @battleblow I decided to not make the change to RUSAGE_THREAD and leave it the way it was. NetBSD doesn't have RUSAGE_THREAD, so if you convert the FreeBSD implementation over to KERN_PROC_PID w/threads as well it can be removed then.
> 
> Here are the relevant tests for testing:
> 
> ```
> gmake run-test-only TEST="hotspot/jtreg/runtime/jni/terminatedThread/TestTerminatedThread.java jdk/jfr/event/runtime/TestThreadCpuTimeEvent.java"
> ```
> 
> okay to merge the current version?

Yes, please go ahead and merge.  I think it is reasonable to leave it as is for now.  Ideally we'd be able to instrument this and see how often each version is called, but I can't think of an easy way to do it.
